File: 0054-spiral-matrix/solution.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) 
    {
        int startR, startC, endR, endC;
        
        startR = startC = 0;
        endR = matrix.size()-1; endC = matrix[0].size()-1;
        
        
        vector<int> result;
        
        while(startR <= endR && startC <= endC)
        {
            
            int r, c;

            // right
            // cout<<"right"<<endl;
            for(r = startR, c = startC; c <= endC; c++)
            {
                // cout<<"("<<r<<","<<c<<") ";
                result.push_back(matrix[r][c]);
            }
            // cout<<endl;
            
            // down
            // cout<<"down"<<endl;
            for(r = startR + 1, c = endC; r <= endR; r++)
            {
                // cout<<"("<<r<<","<<c<<") ";
                result.push_back(matrix[r][c]);
            }
            // cout<<endl;
            
            // left
            // cout<<"left"<<endl;
            for(r = endR, c = endC-1; c >= startC && startR != endR; c--)
            {
                // cout<<"("<<r<<","<<c<<") ";
                result.push_back(matrix[r][c]);
            }
            // cout<<endl;

            // up
            // cout<<"up"<<endl;
            for(r = endR-1, c = startC; r >= startR+1 && startC != endC; r--)
            {
                // cout<<"("<<r<<","<<c<<") ";
                result.push_back(matrix[r][c]);
            }
            // cout<<endl;

            startR++;
            startC++;
            
            endR--;
            endC--;
            
            // cout<<"rows "<<startR<<" "<<endR<<endl;
            // cout<<"cols "<<startC<<" "<<endC<<endl;
        }
        
        return result;
        
    }
};

File: 0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []
        numRows, numCols = len(matrix), len(matrix[0])

        left, right = 0, numCols - 1
        up, down = 0, numRows - 1

        while len(result) != (numRows*numCols):

            for col in range(left, right+1):
                result.append(matrix[up][col])
            
            for row in range(up+1, down+1):
                result.append(matrix[row][right])
            
            if up != down:
                for col in range(right - 1, left - 1, -1):
                    result.append(matrix[down][col])
            
            if left != right:
                for row in range(down - 1, up, -1):
                    result.append(matrix[row][left])

            left += 1
            right -= 1
            up += 1
            down -= 1
        return result
